contracts/tokenfunds.js

```diff
@@ -0,0 +1,583 @@
+    const newDtf = {
+      payToken,
+      voteToken,
+      voteThreshold,
+      maxDays,
+      maxAmountPerDay,
+      proposalFee,
+      active: false,
+      creator: api.sender,
+      lastTickTime: now.getTime(),
+    };
+    newDtf.id = generateDtfId(newDtf);
+    const existingDtf = await api.db.findOne('funds', { id: newDtf.id });
+    if (!api.assert(!existingDtf, 'DTF already exists')) return;
+
+    const insertedDtf = await api.db.insert('funds', newDtf);
+
+    // burn the token creation fees
+    if (api.sender !== api.owner && api.BigNumber(dtfCreationFee).gt(0)) {
+      await api.executeSmartContract('tokens', 'transfer', {
+        // eslint-disable-next-line no-template-curly-in-string
+        to: 'null', symbol: "'${CONSTANTS.UTILITY_TOKEN_SYMBOL}$'", quantity: dtfCreationFee, isSignedWithActiveKey,
+      });
+    }
+    api.emit('createFund', { id: insertedDtf.id });
+  }
+};
+
+actions.updateFund = async (payload) => {
+  const {
+    fundId, voteThreshold, maxDays, maxAmountPerDay, proposalFee, isSignedWithActiveKey,
+  } = payload;
+
+  // get contract params
+  const params = await api.db.findOne('params', {});
+  const { dtfUpdateFee } = params;
+
+  // eslint-disable-next-line no-template-curly-in-string
+  const utilityTokenBalance = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: "'${CONSTANTS.UTILITY_TOKEN_SYMBOL}$'" });
+
+  const authorizedCreation = api.BigNumber(dtfUpdateFee).lte(0) || api.sender === api.owner
+    ? true
+    : utilityTokenBalance && api.BigNumber(utilityTokenBalance.balance).gte(dtfUpdateFee);
+
+  if (api.assert(authorizedCreation, 'you must have enough tokens to cover the creation fee')
+    && api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')
+    && api.assert(typeof voteThreshold === 'string' && api.BigNumber(voteThreshold).gt(0), 'invalid voteThreshold: greater than 0')
+    && api.assert(typeof maxDays === 'string' && api.BigNumber(maxDays).isInteger() && api.BigNumber(maxDays).gt(0) && api.BigNumber(maxDays).lte(730), 'invalid maxDays: integer between 1 and 730')
+    && api.assert(typeof maxAmountPerDay === 'string' && api.BigNumber(maxAmountPerDay).gt(0), 'invalid maxAmountPerDay: greater than 0')) {
+    if (proposalFee) {
+      if (!api.assert(typeof proposalFee === 'object'
+        && typeof proposalFee.method === 'string' && FeeMethod.indexOf(proposalFee.method) !== -1
+        && typeof proposalFee.symbol === 'string'
+        && typeof proposalFee.amount === 'string' && api.BigNumber(proposalFee.amount).gt(0), 'invalid proposalFee')) return;
+      const feeTokenObj = await api.db.findOneInTable('tokens', 'tokens', { symbol: proposalFee.symbol });
+      if (!api.assert(feeTokenObj && api.BigNumber(proposalFee.amount).dp() <= feeTokenObj.precision, 'invalid proposalFee token or precision')) return;
+    }
+    const existingDtf = await api.db.findOne('funds', { id: fundId });
```

Comment:
If the generateDtfId function is required, then should fundId be generated through a similar process so external callers won't have to be concerned with that detail?  Same thing applies to other actions that have a fundId parameter.

Reply:
Inherited from mining, its not required here. Removed.